### ISSUE TITLE
False positive "Unused private elements"

### ISSUE THREAD

**Issue by liepumartins** [nunomaduro/phpinsights#363]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| Library version  | 1.13.0

Following example, methods are just to mimic the behaviour.
```
<?php

class Date
{
    private $datetime;

    public static function getAvailable(): DateTime
    {
        $specific = self::getSpecific();
        if ($specific) {
            return $specific->dateTime();
        }
        return new DateTime();
    }

    private static function getSpecific(): ?Date
    {
        return new Date();
    }

    private function dateTime(): DateTime
    {
        return DateTime::createFromFormat('Y-m-d H:i:s', $this->datetime);
    }
}
```

Insights reports incorrectly, that `dateTime()` is unused:
```
[Code] Unused private elements:
  Date.php:21: Class Date contains unused private method dateTime().
```

**Comment by olivernybroe** [comment]
Thanks for reporting the issue.

We are using [`SlevomatCodingStandard.Classes.UnusedPrivateElements`](https://github.com/slevomat/coding-standard#slevomatcodingstandardclassesunusedprivateelements-) for checking this, so could you post it on their GitHub issues also, as it technically is an error in their sniff. 

However for this project, we might consider if this sniff is too unreliable and disable it, or find another solution for checking for this.

**Comment by liepumartins** [comment]
Looks like this won't be fixed https://github.com/slevomat/coding-standard/issues/905

**Comment by olivernybroe** [comment]
What a shame 😢 
I'll keep this issue open as we might be able to solve it on our side. 
Maybe by replacing it with a phpstan rule if/when we get support for phpstan or some other tool.

**Comment by ondrejmirtes** [comment]
Use PHPStan instead :) Read this blogpost: https://phpstan.org/blog/detecting-unused-private-properties-methods-constants

TL;DR @kukulich decided to deprecate UnusedPrivateElementsSniff in Slevomat Coding Standard, PHPStan should be used for this purpose instead. Make sure you're on 0.12.33 (the latest release) and enable bleedingEdge.neon in your phpstan.neon:

```neon
includes:
	- vendor/phpstan/phpstan/conf/bleedingEdge.neon
```

**Comment by JustSteveKing** [comment]
Keeping open for reference

**Comment by JustSteveKing** [comment]
Looks like dependency investigation is needed for ongoing support, leaning on something like PHPStan over a specific coding standard might be worth investigating

**Comment by jeancodogno** [comment]
In my recent tests with the latest version, this issue has been resolved.
